libs/ledger-live-common/src/deviceSDK/commands/firmwareUpdate/installFirmware.ts

```diff
@@ -0,0 +1,148 @@
+import { Observable, of, throwError } from "rxjs";
+import URL from "url";
+import Transport, { TransportStatusError } from "@ledgerhq/hw-transport";
+import type { FinalFirmware, OsuFirmware } from "@ledgerhq/types-live";
+import type { DeviceInfo, SocketEvent } from "@ledgerhq/types-live";
+import { version as livecommonversion } from "../../../../package.json";
+import { getEnv } from "../../../env";
+//import { createMockSocket, bulkSocketMock, secureChannelMock } from "../../api/socket.mock";
+import { log } from "@ledgerhq/logs";
+import { createDeviceSocket } from "../../../api/socket";
+import { catchError, filter, map } from "rxjs/operators";
+import {
+  ManagerFirmwareNotEnoughSpaceError,
+  UserRefusedFirmwareUpdate,
+  DeviceOnDashboardExpected,
+  ManagerDeviceLockedError,
+} from "@ledgerhq/errors";
+import { UnresponsiveCmdEvent } from "../core";
+
+export type InstallFirmwareCommandRequest = {
+  targetId: DeviceInfo["targetId"];
+  firmware: OsuFirmware | FinalFirmware;
+};
+
+type FilteredSocketEvent =
+  | {
+      type: "bulk-progress";
+      progress: number;
+      index: number;
+      total: number;
+    }
+  | {
+      type: "device-permission-requested";
+      wording: string;
+    };
+
+export type InstallFirmwareCommandEvent =
+  | {
+      type: "progress";
+      progress: number;
+    }
+  | {
+      type: "allowManagerRequested";
+    }
+  | {
+      type: "firmwareInstallPermissionRequested";
+    }
+  | {
+      type: "firmwareInstallPermissionGranted";
+    }
+  | UnresponsiveCmdEvent;
+
+export function installFirmwareCommand(
+  transport: Transport,
+  { targetId, firmware }: InstallFirmwareCommandRequest
+): Observable<InstallFirmwareCommandEvent> {
+  // TODO handle mock
+
+  log("device-command", "installOsuFirmware", {
+    targetId,
+    osuFirmware: firmware,
+  });
+
+  return createDeviceSocket(transport, {
+    url: URL.format({
+      pathname: `${getEnv("BASE_SOCKET_URL")}/install`,
+      query: {
+        targetId,
+        livecommonversion,
+        perso: firmware.perso,
+        firmware: firmware.firmware,
+        firmwareKey: firmware.firmware_key,
+      },
+    }),
+  }).pipe(
+    catchError(remapSocketFirmwareError),
+    filter<SocketEvent, FilteredSocketEvent>((e): e is FilteredSocketEvent => {
+      return (
+        e.type === "bulk-progress" || e.type === "device-permission-requested"
+      );
+    }),
+    map<FilteredSocketEvent, InstallFirmwareCommandEvent>((e) => {
+      if (e.type === "bulk-progress") {
+        return e.index === e.total - 1
```

Comment:
it was probably in the former install firmware logic, but this looks like magic ahah

Reply:
I'm adding some comments explaining